File: app/src/main/java/net/twisterrob/android/view/color/swatches/pixel/color/ColorReplacer.java
package net.twisterrob.android.view.color.swatches.pixel.color;

import net.twisterrob.android.view.color.swatches.pixel.drawer.BitmapDrawer;

public class ColorReplacer implements PixelColor {
    private final PixelColor pixelColor;
    private final int findColor;
    private final int replaceColor;

    public ColorReplacer(PixelColor pixelColor, int findColor, int replaceColor) {
        this.pixelColor = pixelColor;
        this.findColor = findColor;
        this.replaceColor = replaceColor;
    }

    @Override
    public void initializeInvariants(int w, int h) {
        pixelColor.initializeInvariants(w, h);
    }

    @Override
    public int getPixelColorAt(int x, int y) {
        int color = pixelColor.getPixelColorAt(x, y);
        return color != findColor ? color : replaceColor;
    }

    public static BitmapDrawer.Factory wrap(final BitmapDrawer.Factory factory, final int findColor, final int replaceColor) {
        return new BitmapDrawer.Factory() {
            @Override
            public BitmapDrawer create(int[] bitmap, int w, int h, PixelColor color) {
                return factory.create(bitmap, w, h, new ColorReplacer(color, findColor, replaceColor));
            }
        };
    }
}

File: feature/base/src/main/java/net/twisterrob/android/view/color/swatches/pixel/color/ColorReplacer.kt
package net.twisterrob.android.view.color.swatches.pixel.color

import net.twisterrob.android.view.color.swatches.pixel.drawer.BitmapDrawer
import net.twisterrob.android.view.color.swatches.pixel.drawer.BitmapDrawer.Factory

class ColorReplacer(
	private val pixelColor: PixelColor,
	private val findColor: Int,
	private val replaceColor: Int
) : PixelColor by pixelColor {

	override fun getPixelColorAt(x: Int, y: Int): Int {
		val color = pixelColor.getPixelColorAt(x, y)
		return if (color == findColor) replaceColor else color
	}

	companion object {

		@JvmStatic
		fun wrap(factory: Factory, findColor: Int, replaceColor: Int) =
			object : BitmapDrawer.Factory {
				override fun create(bitmap: IntArray, w: Int, h: Int, pixel: PixelColor) =
					factory.create(bitmap, w, h, ColorReplacer(pixel, findColor, replaceColor))
			}
	}
}
